WikiWatershed/model-my-watershed
# Custom Weather: Add Main Workflow

## Overview

A pretty wide-ranging changeset. See commits for details, but mainly test the functionality.

Connects #3287
Connects #3288
Connects #3289
Connects #3290
Connects #3284
Connects #3271 

### Demo

Uploading custom weather data:

![2020-05-12 18 35 47](https://user-images.githubusercontent.com/1430060/81752537-a0ef0d80-947f-11ea-8db3-739a74ff4857.gif)

Switching between custom and default data:

![2020-05-12 18 28 17](https://user-images.githubusercontent.com/1430060/81752583-b5cba100-947f-11ea-90dd-3f12ce31ee1a.gif)

Downloading custom weather CSV:

![2020-05-12 18 30 00](https://user-images.githubusercontent.com/1430060/81752625-d267d900-947f-11ea-90ff-07ecea4dd2a1.gif)

Deleting custom weather:

![2020-05-12 18 30 56](https://user-images.githubusercontent.com/1430060/81752665-eb708a00-947f-11ea-8dff-9ab581cf7a63.gif)

Compare view:

![image](https://user-images.githubusercontent.com/1430060/81752479-83ba3f00-947f-11ea-821e-401ac1f0d7dc.png)

### Notes

Some of the validation for #3284 is done here. More remains.

Will get the proper verbiage from the client for the placeholder text. Need to clarify that uploads must be in CM and C.

There are some lingering edge cases, for which I'll make follow up cards:

* The sidebar title still says "Average monthly water fluxes (in) from 30-years of daily water balance", it should say as many years as the weather data is for
* Need to check before deleting that the weather data is not used in any other scenarios
* If the delete weather data but don't click "Save", but "Cancel", we're in an inconsistent state

## Testing Instructions

* Create a new MapShed project
* Add a new scenario
* Upload a custom weather data file
    - You may use any of these test files: [weather-data-tests.zip](https://github.com/WikiWatershed/model-my-watershed/files/4452949/weather-data-tests.zip)
* Play around and ensure you can:
    - [ ] Upload valid weather data
    - [ ] Be notified of errors for invalid weather data (`missing-headers.csv`)
    - [ ] Download uploaded weather data
    - [ ] Switch between default and custom weather data
    - [ ] Duplicate scenarios and have the weather data carry over
    - [ ] Delete weather data
